### PR TITLE
Add cast related clippy lints

### PR DESCRIPTION
**Pull request by christianpoveda**
# Description of change

This adds some clippy lints to avoid unexpected behavior when using `as` casts. 

## Links to any relevant issues

fixes https://github.com/Alex6323/bee-p/issues/195

## Type of change

- Bug fix (a non-breaking change which fixes an issue)

## How the change has been tested

Describe the tests that you ran to verify your changes.

Make sure to provide instructions for the maintainer as well as any relevant configurations.

## Change checklist

Tick the boxes that are relevant to your changes, and delete any items that are not.

- [x] I have followed the contribution guidelines for this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] (Doesn't apply) I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that new and existing unit tests pass locally with my changes
- [x] (Doesn't apply) I have updated the CHANGELOG.md, if my changes are significant enough

### CODE REVIEW

**Comment by christianpoveda** [comment]
I have a couple questions about warnings caused by the new lints that I think we can fix. I'll add them as a "revieW

**Comment by thibault-martinez** [comment]
Can we create a config file for clippy with the lints in it ? Otherwise we'd have to put the lint in the `lib.rs` of all the crates...

**Comment by christianpoveda** [comment]
> Can we create a config file for clippy with the lints in it ? Otherwise we'd have to put them in the `lib.rs` of all the crates...

sadly no, the `clippy.toml` file is only used to configure lints but not to enable/disable them.

**Comment by thibault-martinez** [comment]
Why only the `bee-crypto` crate though ?

File: bee-crypto/src/ternary/bigint/u384/mod.rs [hunk 1]
```diff
@@ -363,6 +363,7 @@ impl U384<LittleEndian, U32Repr> {
                 accumulator_extent += 1;
             }
 
+            #[allow(clippy::cast_sign_loss)] // This is an slice of `UTrit`s, so `binary_trit` is always positive.
```
Comment (thibault-martinez) [inline bee-crypto/src/ternary/bigint/u384/mod.rs]:
```suggestion
            #[allow(clippy::cast_sign_loss)] // This is a slice of `UTrit`s, so `binary_trit` is always positive.
```

File: bee-crypto/src/ternary/sponge/curlp.rs [hunk 2]
```diff
@@ -64,6 +64,7 @@ impl CurlP {
         #[inline]
         fn truth_table_get(state: &Trits, p: usize, q: usize) -> Btrit {
             // Safe to index `state` since indexes come from iteration on `state`.
+            #[allow(clippy::cast_sign_loss)] // Reason: "Trit's repr is between `-1` and `1`
```
Comment (thibault-martinez) [inline bee-crypto/src/ternary/sponge/curlp.rs]: Only if it is backed by `Btrit` though, may be worth mentioning

**Pull request merged**